### ISSUE TITLE
Filesystem transactions

### ISSUE THREAD

**Issue by martindurant** [dask/dask#5361]
Right now, Dask uses the various implementations of fsspec for writing. Since each task writes one file, the creation of any chunks on a remote server and finalising of the file happen serially in a single thread. However, if there is an exception, then finalising should happen anyway, resulting in partial content (as you would expect when writing to a local file too). This happens upon `del`, so it's timing cannot be guaranteed. 

If there are retries, this should then be over-written, but I can imagine that with the eventual-consistency of S3/GCS, you are not guaranteed as to which version you end up with. The existing transactions mechanism in fsspec could be used to prevent automatically committing whatever was written, but requires some coordination. An additional wrinkle, is that filesystem instances persist, and are shared between threads. 

Dask (and fsspec users in general) would like to have more guarantees about the state of writing, and the ability to commit/discard whole sets of files across a cluster, as well as, ideally, verification that writes/retries did indeed succeed. Also, it might be nice to distribute filesystem configuration globally.

**Comment by martindurant** [comment]
cc @birdsarah @mariusvniekerk @TomAugspurger

**Comment by mrocklin** [comment]
I was just about to cc those same people ... :)

**Comment by mariusvniekerk** [comment]
The other big part is committing the partial writes at roughly the same time.  Most of the filesystems adhere to the s3 block api have a way to stage a file and perform the commit of this elsewhere.

One key part is that a write for a dataset typically targets a directory for where to place output files.  Dask needs an additional component similar to the hadoop OutputCommitter to ensure that these commits of files happen cleanly.  

This component obviously should commit files, but additionally can clean up directories when datasets are overwritten or complaining if there are already files present in the output path.

**Comment by martindurant** [comment]
Yes, I think, that "at the same time" in this context means "when the computation is complete". The transaction mechanism already does that for a single instance, keeping files as uploaded chunks or in a temporary location (depending on backend) and then finalizing/moving everything quasi-atomically.
